oauth2_http/java/com/google/auth/oauth2/CertificateIdentityPoolSubjectTokenSupplier.java

```diff
@@ -0,0 +1,122 @@
+ */
+
+package com.google.auth.oauth2;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonPrimitive;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Paths;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Base64;
+
+/**
+ * Provider for retrieving subject tokens for {@link IdentityPoolCredentials} by reading an X.509
+ * certificate from the filesystem. The certificate file (e.g., PEM or DER encoded) is read, the
+ * leaf certificate is base64-encoded (DER format), wrapped in a JSON array, and used as the subject
+ * token for STS exchange.
+ */
+public class CertificateIdentityPoolSubjectTokenSupplier
+    implements IdentityPoolSubjectTokenSupplier {
+
+  private static final Gson GSON = new Gson();
+  private final IdentityPoolCredentialSource credentialSource;
+
+  CertificateIdentityPoolSubjectTokenSupplier(IdentityPoolCredentialSource credentialSource) {
+    this.credentialSource = checkNotNull(credentialSource, "credentialSource cannot be null");
+    // This check ensures that the credential source was intended for certificate usage.
+    // IdentityPoolCredentials logic should guarantee credentialLocation is set in this case.
+    checkNotNull(
+        credentialSource.certificateConfig,
+        "credentialSource.certificateConfig cannot be null when creating"
+            + " CertificateIdentityPoolSubjectTokenSupplier");
+  }
+
+  private static X509Certificate loadLeafCertificate(String path)
+      throws IOException, CertificateException {
+    byte[] leafCertBytes;
+    try {
+      // IdentityPoolCredentials should have already validated the path exists via X509Provider.
+      leafCertBytes = Files.readAllBytes(Paths.get(path));
+    } catch (InvalidPathException e) {
+      throw new IOException("Invalid certificate file path provided: " + path, e);
+    }
+    // Files.readAllBytes throws IOException for other read errors.
+    return parseCertificate(leafCertBytes);
+  }
+
+  private static X509Certificate parseCertificate(byte[] certData) throws CertificateException {
+    if (certData == null || certData.length == 0) {
+      throw new IllegalArgumentException("Invalid certificate data: empty or null input");
+    }
+
+    try {
+      CertificateFactory certificateFactory = CertificateFactory.getInstance("X.509");
+      InputStream certificateStream = new ByteArrayInputStream(certData);
+      return (X509Certificate) certificateFactory.generateCertificate(certificateStream);
+    } catch (CertificateException e) {
+      throw new CertificateException("Failed to parse X.509 certificate data.", e);
```

Comment:
qq, why is this re-thrown? Is this for better error messages?

Reply:
Yes, the main reason for re-throwing the CertificateException here is to provide a more specific error message ("Failed to **parse** X.509 certificate data.") that adds context about what operation was being attempted when the failure occurred.

Reply:
Sg, thanks! Could you add a small comment regarding adding more info into the exception?